nx/lib/nx.ex

```diff
@@ -7919,6 +7921,101 @@ defmodule Nx do
     impl!(a, b).triangular_solve(%{b | type: output_type}, a, b, [])
   end
 
+  @doc """
+  Invert a square 2-D tensor.
+
+  For non-square tensors, use `Nx.svd/2` for pseudo-inverse calculations.
+
+  ## Examples
+
+      iex> a = Nx.tensor([[1, 0, 0, 0], [2, 1, 0, 0], [1, 0, 1, 0], [1, 1, 1, 1]])
+      iex> a_inv = Nx.invert(a)
+      #Nx.Tensor<
+        f32[4][4]
+        [
+          [1.0, 0.0, 0.0, 0.0],
+          [-2.0, 1.0, 0.0, 0.0],
+          [-1.0, 0.0, 1.0, 0.0],
+          [2.0, -1.0, -1.0, 1.0]
+        ]
+      >
+      iex> Nx.dot(a, a_inv)
+      #Nx.Tensor<
+        f32[4][4]
+        [
+          [1.0, 0.0, 0.0, 0.0],
+          [0.0, 1.0, 0.0, 0.0],
+          [0.0, 0.0, 1.0, 0.0],
+          [0.0, 0.0, 0.0, 1.0]
+        ]
+      >
+      iex> Nx.dot(a_inv, a)
+      #Nx.Tensor<
+        f32[4][4]
+        [
+          [1.0, 0.0, 0.0, 0.0],
+          [0.0, 1.0, 0.0, 0.0],
+          [0.0, 0.0, 1.0, 0.0],
+          [0.0, 0.0, 0.0, 1.0]
+        ]
+      >
+
+  ### Error cases
+
+      iex> Nx.invert(Nx.tensor([[3, 0, 0, 0], [2, 1, 0, 0]]))
+      ** (ArgumentError) can only invert square matrices, got: {2, 4}
+
+      iex> Nx.invert(Nx.tensor([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [1, 1, 1, 1]]))
+      ** (ArgumentError) can't invert a singular matrix
+
+  """
+  @doc type: :linalg
+  def invert(tensor) do
+    %T{shape: {m, n}} = tensor = tensor!(tensor)
+
+    if m != n,
+      do:
+        raise(
+          ArgumentError,
+          "can only invert square matrices, got: {#{m}, #{n}}"
+        )
+
+    # We need to achieve an LQ decomposition for `tensor` (henceforth called A)
+    # because triangular_solve only accepts lower_triangular matrices
+    # Therefore, we can use the fact that if we have B = Q'R' -> transpose(B) = LQ.
+    # If we set B = transpose(A), we can get A = LQ through transpose(qr(transpose(A)))
+
+    # Given A = LQ, we can then solve Ax = b as shown below
+    # Ax = b -> L(Qx) = b -> Ly = b, where y = Qx -> x = transpose(Q)b
+
+    # Finally, it can be shown that inv(A) can be calculated by setting
+    # b as each of the I matrix columns, which means that if we can solve Ax = b
+    # we can invert A by iteratively calculating x
+
+    {qt, r_prime} = tensor |> transpose() |> qr()
+
+    l = transpose(r_prime)
+
+    identity = eye(n)
+
+    n
+    |> count_up(0)
+    |> Enum.map(fn i ->
+      b = identity |> slice([0, i], [n, 1]) |> Nx.reshape({n})
+
+      y =
+        try do
+          triangular_solve(l, b)
+        rescue
+          ArgumentError ->
+            raise ArgumentError, "can't invert a singular matrix"
+        end
```

Comment:
This try/catch won't work when invoking inverse inside a `defn` so perhaps we should remove it? :) What is its purpose?

Reply:
Better error messages. Triangular solve will return something like "can't solve singular system", which isn't the most semantic message, but this isn't necessary for the algorithm itself

Reply:
Ok. The issue is that this error message won't work inside `defn`, so perhaps we should improve the error message in the BinaryBackend if possible.

Reply:
@josevalim Just for curiosity, what are current limitations of `defn` and what do we expect to have solved in terms of limitations by the time the API becomes stable?